packages/labs/compiler/src/lib/ast-fragments.ts

```diff
@@ -7,14 +7,155 @@
-  index: number;
+export type TemplatePart =
+  | {
+      type:
+        | typeof PartType.CHILD
+        | typeof PartType.COMMENT_PART
+        | typeof PartType.ELEMENT;
+      index: number;
+    }
+  | {
+      type: typeof PartType.ATTRIBUTE;
+      index: number;
+      name: string;
+      strings: Array<string>;
+      ctorType: PartType;
+    };
+
+const attributePartConstructors = {
+  [PartType.ATTRIBUTE]: 'AttributePart',
+  [PartType.PROPERTY]: 'PropertyPart',
+  [PartType.BOOLEAN_ATTRIBUTE]: 'BooleanAttributePart',
+  [PartType.EVENT]: 'EventPart',
+} as const;
+
+// These constructors have been renamed to reduce the chance of a naming collision.
+export interface AttributePartConstructorAliases {
+  AttributePart?: ts.Identifier;
+  PropertyPart?: ts.Identifier;
+  BooleanAttributePart?: ts.Identifier;
+  EventPart?: ts.Identifier;
+}
+
+/**
+ * Creates the statement:
+ *
+ * ```ts
+ * import * as <namespace> from "<moduleSpecifier>"
+ * ```
+ */
+const importNamespaceDeclaration = (
+  factory: ts.NodeFactory,
+  namespace: ts.Identifier,
+  moduleSpecifier: ts.StringLiteral
+) =>
+  factory.createImportDeclaration(
+    undefined,
+    factory.createImportClause(
+      false,
+      undefined,
+      factory.createNamespaceImport(namespace)
+    ),
+    moduleSpecifier,
+    undefined
+  );
+
+/**
+ * Add Parts constructors import. To avoid identifier collisions, each part
+ * constructor is mapped to a unique identifier that must be provided. If no
+ * identifier is provided for a part, then the part will not be emitted.
+ *
+ * This uses a namespace import for g3 compatibility.
+ *
+ * ```ts
+ * import * as litHtmlPrivate_1 from "lit-html/private-ssr-support.js";
+ * const { AttributePart: <AttributePart alias>, PropertyPart: <PropertyPart alias>, BooleanAttributePart: <BooleanAttributePart alias>, EventPart: <EventPart alias> } = litHtmlPrivate_1._$LH;
+ * ```
+ */
+export const addPartConstructorImport = ({
+  factory,
+  sourceFile,
+  securityBrand,
+  partIdentifiers,
+}: {
+  sourceFile: ts.SourceFile;
+  securityBrand: ts.Statement;
+  factory: ts.NodeFactory;
+  partIdentifiers: AttributePartConstructorAliases;
+}): ts.SourceFile => {
+  const uniqueLitHtmlPrivateIdentifier =
+    factory.createUniqueName('litHtmlPrivate');
+  const brandIdx = sourceFile.statements.indexOf(securityBrand);
+  if (brandIdx === -1) {
+    throw new Error(
+      `Internal Error: Could not find security brand declaration.`
+    );
+  }
+  // Add parts import immediately before the security brand.
+  const beforeSecurityBrand = sourceFile.statements.slice(0, brandIdx);
+  const afterSecurtyBrand = sourceFile.statements.slice(brandIdx);
+
+  const partConstructors = [
+    'AttributePart',
+    'PropertyPart',
+    'BooleanAttributePart',
+    'EventPart',
+  ] as const;
+  const partsObjectBinding = partConstructors
+    .map((part) => {
+      const identifierAlias = partIdentifiers[part];
+      if (!identifierAlias) {
```

Comment:
If I'm following this, you allow an object to be provided that renames the part constructors that are imported.

Who decides on the mapping, and is this needed? Can't you create a new unique name for the opt-level scope with TS APIs(`factory.createUniqueName`) and always rename the constructors?

Reply:
You are following correctly. The prep stage is where we decide which parts to import (because we know which parts have been encountered). Therefore if a file contains no attribute parts, we do not need to generate any imports.

The unique names are to provide defense against naming collisions – in the edge case where a file already contains an `AttributePart` identifier.

Reply:
Can't you just use `factory.createUniqueName` instead of taking a mapping?

Reply:
Currently the order of operations, taken from `CompiledTemplatePass.getTransformer()` are:
 1. Do a traversal of the AST to locate all the templates. `pass.findTemplates`
 2. Rewrite the found templates. `pass.rewriteTemplates`. If an attribute part requires a constructor, the unique name is created here, and added to the part constructor mapping.
 3. Finally, if any part constructors are used, add the part constructor import to the source file.

The benefit of this ordering is that we only add the import if a compiled template needs the constructors. Maybe I've overcomplicated it.

An alternative order could be. 

1. Generate all the unique constructor identifier up front.
2. Do steps 1. and 2. above. While rewriting the templates track which constructors are actually used.
3. Then pass in a set of which ones we want to actually add to the file.

But I think the book keeping is still required.

Another alternative is to not bother doing any book–keeping, and add the import and all the Attribute part constructors regardless. Although that would require a downstream transform to tree shake or dead code eliminate them.

Reply:
Ah, and going back to your top questions.

> Who decides on the mapping, and is this needed?
The `rewriteTemplate` method only adds a `uniqueIdentifier` for an attribute part constructor if it is used. It's not necessarily needed. We could do one of the alternatives in my comment above.

> Can't you create a new unique name for the opt-level scope with TS APIs(factory.createUniqueName) and always rename the constructors.

What is the `opt-level` scope? In the current implementation the constructors are always renamed. They're just instantiated with `factory.createUniqueName` during the `rewriteTemplate` pass; so we both track the unique name and which ones are used in a `CompiledTemplate`.  Then we only import and pull the constructors off `_$LH` that are used.

Reply:
Sorry, I meant "top-level" scope - ie, a name unique in that scope. But I mispoke a bit because I'm not sure that the TS API has scope-aware unique name generation, and the name has to be unique in all scopes that have templates, not just the top-level scope. Anyway, `factory.createUniqueName()` is fine.